applications/GeoMechanicsApplication/tests/cpp_tests/test_geomechanics_erosion_piping.cpp

```diff
@@ -0,0 +1,215 @@
+
+namespace Kratos
+{
+template <class TSparseSpace, class TDenseSpace, class TLinearSolver>
+class MockGeoMechanicsNewtonRaphsonErosionProcessStrategy
+    : public GeoMechanicsNewtonRaphsonErosionProcessStrategy<TSparseSpace, TDenseSpace, TLinearSolver>
+{
+public:
+    using BaseType = ImplicitSolvingStrategy<TSparseSpace, TDenseSpace, TLinearSolver>;
+    using TConvergenceCriteriaType = ConvergenceCriteria<TSparseSpace, TDenseSpace>;
+    using TBuilderAndSolverType    = typename BaseType::TBuilderAndSolverType;
+    using TSchemeType              = typename BaseType::TSchemeType;
+
+    MockGeoMechanicsNewtonRaphsonErosionProcessStrategy(ModelPart&                    rModelPart,
+                                                        typename TSchemeType::Pointer pScheme,
+                                                        typename TLinearSolver::Pointer pNewLinearSolver,
+                                                        typename TConvergenceCriteriaType::Pointer pNewConvergenceCriteria,
+                                                        typename TBuilderAndSolverType::Pointer pNewBuilderAndSolver,
+                                                        Parameters& rParameters)
+        : GeoMechanicsNewtonRaphsonErosionProcessStrategy<TSparseSpace, TDenseSpace, TLinearSolver>(
+              rModelPart, pScheme, pNewLinearSolver, pNewConvergenceCriteria, pNewBuilderAndSolver, rParameters),
+          mrModelPart(rModelPart) {};
+
+private:
+    bool Recalculate() override
+    {
+        // The function provides a pressure change over iterations that leads to a non-linear growth of an equilibrium pipe height.
+        // As a result the height curve crosses twice a pipe height growth curve in the search algorithm implemented
+        // in  GeoMechanicsNewtonRaphsonErosionProcessStrategy. The search stops at the second point.
+        for (auto& node : mrModelPart.Nodes()) {
+            node.FastGetSolutionStepValue(WATER_PRESSURE) =
+                pow(node.FastGetSolutionStepValue(WATER_PRESSURE), 0.9725);
+        }
+        return true;
+    }
+
+    void BaseClassFinalizeSolutionStep() override {
+        // to avoid calling the BaseClassFinalizeSolutionStep in tested class.
+    };
+    ModelPart& mrModelPart;
+};
+
+Node::Pointer CreateNodeWithFastSolutionStepValues(
+    ModelPart& rModelPart, int Id, double CoordinateX, double CoordinateY, double WaterPressure)
+{
+    constexpr double coordinate_z = 0.0;
+    auto             p_node = rModelPart.CreateNewNode(Id, CoordinateX, CoordinateY, coordinate_z);
+
+    p_node->FastGetSolutionStepValue(WATER_PRESSURE) = WaterPressure;
+    const array_1d<double, 3> GravitationalAcceleration{0, -9.81, 0};
+    p_node->FastGetSolutionStepValue(VOLUME_ACCELERATION) = GravitationalAcceleration;
+
+    return p_node;
+}
+
+Geometry<Node>::Pointer CreateQuadrilateral2D4N(ModelPart&       rModelPart,
+                                                std::vector<int> NodeIds,
```

Comment:
I would suggest to pass `NodeIds` by reference-to-const rather than by value, to avoid unnecessary copies.
```suggestion
                                                const std::vector<int>& NodeIds,
```

Please, be aware that when you make this change, you also need to rename the function parameter to comply with the Kratos Style Guide: `NodeIds` => `rNodeIds`.

Reply:
Changed